ukbb_qc/release/prepare_vcf_data_release.py

```diff
@@ -601,121 +650,222 @@ def main(args):
+            logger.info("Determining partitioning for %s", contig)
+            chrom_var_map = args.chrom_var_map
+            n_var_per_shard = args.n_var_per_shard
 
-                logger.info("Adjusting partitions...")
-                mt = mt.naive_coalesce(args.n_shards)
-                ht = mt.rows()
-                # Unkey HT to avoid this error with map_partitions:
-                # ValueError: Table._map_partitions must preserve key fields
-                ht = ht.key_by()
-
-                hl.export_vcf(
-                    mt,
-                    release_vcf_path(*tranche_data, contig=contig),
-                    metadata=header_dict,
-                    append_to_header=append_to_vcf_header_path(*tranche_data),
-                    tabix=True,
-                )
+            # Round to the nearest whole number
+            n_partitions = round(chrom_var_map[contig] / n_var_per_shard)
 
-                logger.info("Getting start and stops per shard...")
+            # Round the number of partitions to the relevant place
+            # e.g., return 4000 instead of 3839, 900 instead of 913, 70 instead of 69
+            # NOTE: The number of shards isn't guaranteed to be even
+            # `hl.naive_coalesce` will return the number of shards found if smaller than desired number
+            # e.g., if the desired number is 2000, but the chromosome only has 1789 partitions,
+            # `hl.naive_coalesce` will do nothing and the chromosome will export to 1789 shards
+            n_digits = -(len(str((n_partitions))) - 1)
+            n_partitions = round(n_partitions, n_digits)
 
-                def part_min_and_max(part):
-                    keys = part.map(lambda x: x.select("locus", "alleles"))
-                    return hl.struct(start=keys[0], end=keys[-1])
+            # Read in MT and filter to contig
+            # Repartition to a large number of partitions here so that the chromosomes have closer to the desired number of shards
+            mt = hl.read_matrix_table(
+                "gs://broad-ukbb/broad.freeze_7/release/ht/broad.freeze_7.release.vcf.ukb_official_export.mt",
+                _n_partitions=40000,
+            ).select_rows()
 
-                print(
-                    ht._map_partitions(
-                        lambda p: hl.array([part_min_and_max(p)])
-                    ).collect()
-                )
+            if args.test:
+                mt = mt._filter_partitions(range(2))
+            mt = hl.filter_intervals(mt, [hl.parse_locus_interval(contig)])
+            mt = mt.annotate_rows(**ht[mt.row_key])
+
+            logger.info("Adjusting partitions...")
+            mt = mt.naive_coalesce(n_partitions)
+            logger.info("%s has %i partitions", contig, mt.n_partitions())
+
+            logger.info("Densifying...")
+            mt = hl.experimental.densify(mt)
+            # Drop END and het non ref to avoid exporting
+            mt = mt.drop("het_non_ref")
```

Comment:
Did you mean to drop END annotations in this line as well?

Reply:
I dropped END here initially but forgot that annotation gets dropped after densify
```
INFO (vcf_release 740): Densifying and exporting VCF...
INFO (vcf_release 790): Copying hail log to logging bucket...
2021-10-27 21:31:35 Hail: INFO: copying log to 'gs://broad-ukbb/broad.freeze_7/temp/logs/'...
Traceback (most recent call last):
  File "/tmp/2f4ef21da25a44f3b1edbb462955650d/prepare_vcf_data_release.py", line 855, in <module>
    main(args)
  File "/tmp/2f4ef21da25a44f3b1edbb462955650d/prepare_vcf_data_release.py", line 743, in main
    mt = mt.drop("END", "het_non_ref")
  File "<decorator-gen-1211>", line 2, in drop
  File "/opt/conda/default/lib/python3.8/site-packages/hail/typecheck/check.py", line 577, in wrapper
    return __original_func(*args_, **kwargs_)
  File "/opt/conda/default/lib/python3.8/site-packages/hail/matrixtable.py", line 1330, in drop
    raise IndexError("MatrixTable has no field '{}'".format(e))
IndexError: MatrixTable has no field 'END'
```
https://console.cloud.google.com/dataproc/jobs/2f4ef21da25a44f3b1edbb462955650d/monitoring?region=us-central1&project=maclab-ukbb